src/pact.erl

```diff
@@ -1,100 +1,121 @@
-
-cleanup_pact(PactRef) ->
-    pact_ffi_nif:erl_pactffi_free_pact_handle(PactRef).
-
-with_query_parameter(InteractionRef, Name, Index, Value) ->
-    pact_ffi_nif:erl_pactffi_with_query_parameter_v2(InteractionRef, Name, Index, Value).
-
-given(InteractionRef, State) ->
-    pact_ffi_nif:erl_pactffi_given(InteractionRef, State).
+create_interaction(PactPid, Interaction) ->
+    {Consumer, Producer} = pact_handler:get_consumer_producer(PactPid),
+    PactRef = pact_ffi_interface:create_new_pact(Consumer, Producer),
+    ok = pact_handler:set_pact_ref(PactPid, PactRef),
+    GivenState = maps:get(upon_receiving, Interaction, <<"">>),
+    InteractionRef = pact_ffi_interface:create_new_interaction(PactRef, GivenState),
+    ok = pact_handler:create_interaction(PactPid, InteractionRef, Interaction),
+    RequestDetails = maps:get(with_request, Interaction, #{}),
+    ok = insert_request_details(InteractionRef, RequestDetails),
+    ResponseDetails = maps:get(will_respond_with, Interaction ,#{}),
+    ok = insert_response_details(InteractionRef, ResponseDetails),
+    MockServerPort = pact_ffi_interface:create_mock_server_for_transport(
+        PactRef, <<"127.0.0.1">>, 0, <<"http">>
+    ),
+    ok = pact_handler:set_mock_server_port(PactPid, MockServerPort),
+    {ok, MockServerPort}.
+
+verify_interaction(PactPid) ->
+    MockServerPort = pact_handler:get_mock_server_port(PactPid),
+    {ok, matched} = pact_ffi_interface:verify(MockServerPort).
+
+verify_interaction_and_write_pact(PactPid, Path) ->
+    verify_interaction(PactPid),
+    write_pact_file(PactPid, Path).
+
+write_pact_file(PactPid, Path) ->
+    PactRef = pact_handler:get_pact_ref(PactPid),
+    pact_ffi_interface:write_pact_file(PactRef, Path),
+    cleanup_internal(PactPid).
+
+cleanup_internal(PactPid) ->
+    PactRef = pact_handler:get_pact_ref(PactPid),
+    MockServerPort = pact_handler:get_mock_server_port(PactPid),
+    ok = pact_ffi_interface:cleanup_mock_server(MockServerPort),
+    pact_ffi_interface:cleanup_pact(PactRef).
+
+cleanup_pact(PactPid) ->
+    pact_handler:stop(PactPid).
+
+insert_request_details(InteractionRef, RequestDetails) ->
+    ReqMethod = maps:get(method, RequestDetails),
+    ReqPath = maps:get(path, RequestDetails),
+    pact_ffi_interface:with_request(InteractionRef, ReqMethod, ReqPath),
+    ReqHeaders = maps:get(headers, RequestDetails, undefined),
+    case ReqHeaders of
+        undefined -> ok;
+        _ ->
+            maps:fold(
+                fun(Key, Value, _Acc) ->
+                    pact_ffi_interface:with_request_header(InteractionRef, Key, 0, Value)
+                end,
+                ok,
+                ReqHeaders
+            )
+    end,
+    ReqBody = maps:get(body, RequestDetails, undefined),
+    case ReqBody of
+        undefined -> ok;
+        _ ->
+            Body = maps:get(body, ReqBody, <<"">>),
+            ContentType = maps:get(content_type, ReqBody, <<"">>),
```

Comment:
Why 2-level deep body? content_type can be at RequestDetails level

Reply:
I kept the body json and its content type together so that it aligns with the pact ffi function that takes json and content type together. Maybe i should rename the inner body to body_content ?

Reply:
Since this might be nested even more when matchers are introduced, I would like to keep content_type optional and outside the body in the API exposed to the developer.

The interaction map can have required and optional keys. Maybe something like:

```erlang
#{
  upon_receiving => <<"a request to write to influxdb">>,
  with_request => #{
    method => <<"POST">>,
    path => <<"/write?db=mydb">>,
    headers => #{<<"key">> => <<"value">>},               %% optional
    content_type => <<"application/octet-stream">>        %% optional
    body => <<"cpu_load_short,host=server01 value=0.64">> %% optional
    body_json => #{                                       %% optional
      id => pact_matchers.integer(123),
      status => pact_matchers.regex(<<"created|completed">>)
    }
}
```

This will provide an easy API for developers, while the pact_erlang code will call the right FFI function with args based on what is passed.

Reply:
See also the ResponseBuilder class in PactNet: https://github.com/pact-foundation/pact-net/blob/master/src/PactNet/ResponseBuilder.cs#L276

Reply:
Now fetching Content-Type from headers in request and response, please check